File: utilities/strings/java/src/main/java/io/polyglotutilities/strings/StringUtils.java
package io.polyglotutilities.strings;

import java.util.Locale;
import java.util.regex.Pattern;

public final class StringUtils {
    private static final Pattern EMAIL_PATTERN = Pattern.compile("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$");

    private StringUtils() {
    }

    public static boolean isEmpty(String value) {
        return value == null || value.isEmpty();
    }

    public static String capitalize(String value) {
        if (value == null || value.isEmpty()) {
            return value;
        }
        int firstCodePoint = value.codePointAt(0);
        return new String(Character.toChars(Character.toUpperCase(firstCodePoint))) + value.substring(Character.charCount(firstCodePoint));
    }

    public static String reverseString(String value) {
        if (value == null) {
            return null;
        }
        return value.codePoints()
                .mapToObj(codePoint -> new String(Character.toChars(codePoint)))
                .reduce("", (result, character) -> character + result);
    }

    public static boolean isValidEmail(String value) {
        return value != null && EMAIL_PATTERN.matcher(value).matches();
    }

    public static String normalizeWhitespace(String value) {
        return value == null ? null : value.trim().replaceAll("\\s+", " ");
    }

    public static String toTitleCase(String value) {
        if (value == null) {
            return null;
        }
        String normalized = normalizeWhitespace(value).toLowerCase(Locale.ROOT);
        StringBuilder result = new StringBuilder();
        boolean capitalizeNext = true;
        for (int i = 0; i < normalized.length(); i++) {
            char character = normalized.charAt(i);
            if (Character.isWhitespace(character)) {
                result.append(character);
                capitalizeNext = true;
            } else if (capitalizeNext) {
                result.append(Character.toUpperCase(character));
                capitalizeNext = false;
            } else {
                result.append(character);
            }
        }
        return result.toString();
    }
}

File: utilities/strings/kotlin/src/main/kotlin/io/polyglotutilities/strings/StringUtils.kt
package io.polyglotutilities.strings

import java.util.Locale

object StringUtils {
    private val emailPattern = Regex("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$")

    fun isEmpty(value: String?): Boolean = value.isNullOrEmpty()

    fun capitalize(value: String?): String? {
        if (value.isNullOrEmpty()) {
            return value
        }

        val first = value.codePointAt(0)
        val firstString = String(Character.toChars(first)).uppercase(Locale.ROOT)
        return firstString + value.substring(Character.charCount(first))
    }

    fun reverseString(value: String?): String? {
        if (value == null) {
            return null
        }

        val codePoints = value.codePoints().toArray()
        val builder = StringBuilder(value.length)
        for (index in codePoints.indices.reversed()) {
            builder.appendCodePoint(codePoints[index])
        }
        return builder.toString()
    }

    fun isValidEmail(value: String?): Boolean =
        value != null && emailPattern.matches(value)

    fun normalizeWhitespace(value: String?): String? =
        value?.trim()?.replace(Regex("\\s+"), " ")

    fun toTitleCase(value: String?): String? {
        if (value == null) {
            return null
        }

        val normalized = normalizeWhitespace(value)!!.lowercase(Locale.ROOT)
        val words = normalized.split(' ')
        return words.joinToString(" ") { word ->
            if (word.isEmpty()) word
            else word.substring(0, 1).uppercase(Locale.ROOT) + word.substring(1)
        }
    }
}
